src/doc/trpl/error-handling.md

```diff
@@ -1,308 +1,2147 @@
+    }
+}
 ```
 
+This is a little better, but now we've written a lot more code! The case
+analysis has once again bitten us.
 
-This is appropriate for the simplest of cases, but doesn't give us a lot of
-information in the failure case. What if we wanted to know _why_ the function
-failed? For this, we can use the `Result<T, E>` type. It looks like this:
+Combinators to the rescue! Just like `Option`, `Result` has lots of combinators
+defined as methods. There is a large intersection of common combinators between
+`Result` and `Option`. In particular, `map` is part of that intersection:
 
 ```rust
-enum Result<T, E> {
-   Ok(T),
-   Err(E)
+use std::num::ParseIntError;
+
+fn double_number(number_str: &str) -> Result<i32, ParseIntError> {
+    number_str.parse::<i32>().map(|n| 2 * n)
+}
+
+fn main() {
+    match double_number("10") {
+        Ok(n) => assert_eq!(n, 20),
+        Err(err) => println!("Error: {:?}", err),
+    }
 }
 ```
 
-This enum is provided by Rust itself, so you don't need to define it to use it
-in your code. The `Ok(T)` variant represents a success, and the `Err(E)` variant
-represents a failure. Returning a `Result` instead of an `Option` is recommended
-for all but the most trivial of situations.
+The usual suspects are all there for `Result`, including
+[`unwrap_or`](http://doc.rust-lang.org/std/result/enum.Result.html#method.unwrap_or)
+and
+[`and_then`](http://doc.rust-lang.org/std/result/enum.Result.html#method.and_then).
+Additionally, since `Result` has a second type parameter, there are combinators
+that affect only the error type, such as
+[`map_err`](http://doc.rust-lang.org/std/result/enum.Result.html#method.map_err)
+(instead of `map`) and
+[`or_else`](http://doc.rust-lang.org/std/result/enum.Result.html#method.or_else)
+(instead of `and_then`).
 
-Here's an example of using `Result`:
+
+#### The `Result` type alias idiom
+
+In the standard library, you may frequently see types like `Result<i32>`. But
+wait,
+[we defined `Result`](#code-result-def-1)
+to have two type parameters. How can we get away with only specifying one? The
+key is to define a `Result` type alias that *fixes* one of the type parameters
+to a particular type. Usually the fixed type is the error type. For example,
+our previous example parsing integers could be rewritten like this:
 
 ```rust
-#[derive(Debug)]
-enum Version { Version1, Version2 }
+use std::num::ParseIntError;
+use std::result;
+
+type Result<T> = result::Result<T, ParseIntError>;
+
+fn double_number(number_str: &str) -> Result<i32> {
+    unimplemented!();
+}
+```
+
+Why would we do this? Well, if we have a lot of functions that could return
+`ParseIntError`, then it's much more convenient to define an alias that always
+uses `ParseIntError` so that we don't have to write it out all the time.
+
+The most prominent place this idiom is used in the standard library is with
+[`io::Result`](http://doc.rust-lang.org/std/io/type.Result.html). Typically,
```

Comment:
okay I thought there weren't so many of these, but there are, so I'm just gonna stop :)